### ISSUE TITLE
Istio controller must process only k8s ingress resources annotated with "istio"

### ISSUE THREAD

**Issue by andraxylia** [istio/manager#371]
This is to allow co-existence of other ingress controller in a cluster.

**Comment by ijsnellf** [comment]
@andraxylia: it looks like the NGINX and GCE ingress controllers consider non-annotated ingress resources fair game:

NGINX:
https://github.com/kubernetes/ingress/tree/master/controllers/nginx#running-multiple-ingress-controllers

GCE:
https://github.com/kubernetes/ingress/blob/116fbe8c33b5d51b825038d4c7f2a481ab7cee74/controllers/gce/controller/utils.go#L115

**Comment by kyessenov** [comment]
This is a known problem with ingress controller design in k8s. Doing what other k8s controllers do seems reasonable.

**Comment by zcahana** [comment]
We can support the two behaviors with a flag, e.g.:
```
--ingress_class="istio"            // specifies the ingress class annotation to match
--ingress_class_strict=false    // specifies whether only annotated ingress resources should be processed
```
With non-strict being the common behavior for ingress controllers, and strict being the desirable one (IMO).

**Comment by andraxylia** [comment]
The ingress controller processing the ingress resource, assigns it an IP:port equal to the <ingress-controller-node>:<ingress-controller-port>:
loadBalancer:
    ingress:
    - ip: 35.186.247.88

$ kubectl get ing
NAME            HOSTS     ADDRESS         PORTS     AGE
basic-ingress   *         35.186.247.88   80        18h

This is the address where traffic coming from outside of the cluster should be sent. Traffic should not be sent directly to the ingress controller service, like it is currently the case with the bookinfo demo. The user is not supposed to know about the ingress controller service name or other internal details.

With multiple ingress controllers in the same cluster, we can annotate the ingress with "kubernetes.io/ingress.class" to ensure a single controller processes the ingress. However, if the annotation is not set, and more then one controller processes the ingress and sets the IP address, we end up with a race condition.

Note that, unlike nginx,  traefik controller does behave, there is long thread here on the same topic:
https://github.com/containous/traefik/issues/1163

I checked with Tim Hockin and the behavior of the nginx controller is incorrect, and it has to be fixed. Each controller has to know for which class it is responsible. One of the controller has to be the default.

To me, it makes sense to have the default controller be the cloud LB. For deployments in public cloud, this relieves the user from knowing the type of cloud LB, and eases portability between clouds, since no change will be needed in the yaml files.

I re-open this issue to fix three things:
1) Istio default behavior should be to only process istio annotated ingress resources. The extra config flag can be enabled to allow istio to process non-annotated resources. Since this is related to the fact that istio is the default LB, I propose this to be renamed --default_lb=true. This is easier to understand than "ingress-class_strict" or "parse_annotated".

2) Assign an IP address:port to the ingress resource

3) Fix the bookinfo demo to use ingress resource IP and not the external IP of the istio-ingress-controller service

4)optional: Rename istio-ingress-controller to istio-ingress-lb. In fact, that container runs both management & data plane. The mgmt plane is build with a control loop, similar with k8s controller, and it is a k8s ingress controller, no doubt,  hence the name.  But it is in general an HTTP LB, and traffic coming from outside the cluster does pass through the envoy proxy/data plane running inside istio-ingress-controller.

**Comment by zcahana** [comment]
@andraxylia 

1. I'll add the `--default_ingress_controller=true` and `--ingress_class="istio"` flags. Note the flag rename, as explained in (4) below.
2. This will be covered in #214.
3. Depends on #214, therefore not part of this issue.
4. Given that the container runs both management and data plane functions, but can have only a single name, I propose we stick to "ingress controller", so that we're consistent with kubernetes terminology. IMO, "ingress lb" mixes up two distinct (although related) kubernetes concepts, which might lead to confusion.

**Comment by kyessenov** [comment]
can we close it?

**Comment by andraxylia** [comment]
We can close it, since @zcahana will address all the mentioned points as part of #214 .
